File: java/src/main/java/com/example/cote/PGMRS/사칙연산.java
package com.example.cote.PGMRS;

import java.util.*;
import java.nio.file.*;
import java.io.*;

public class 사칙연산 {
	private static final int MIN_ARRAY_LENGTH =  3;
	private static final int MAX_ARRAY_LENGTH =  201;
	private static final int MIN_NUMBER_SIZE = 1000;

	private static interface IntComparator extends Comparator<Integer> {}
	private static final IntComparator[] COMP = {
		(a,b) -> Integer.compare(a,b),
		(a,b) -> Integer.compare(b,a)
	};

	private static final int[] TARGET_VALUE = {
		Integer.MIN_VALUE,
		Integer.MAX_VALUE
	};

	private static final int[][][] cache = new int[MAX_ARRAY_LENGTH+1][MAX_ARRAY_LENGTH+1][2];

	private static int compute(int start, int end, int type, String[] tokens){
		
		int ret = cache[start][end][type];
		if( ret != Integer.MIN_VALUE){
			return ret;
		}

		if( end - start == 1) {
			return Integer.parseInt(tokens[start]);
		}

		ret = TARGET_VALUE[type];

		for(int i = start +1; i< end; i+=2){
			int l = compute(start, i, type, tokens);
			int v; 

			if(tokens[i].equals("+")){
				int r= compute(i+1, end, type, tokens);
				v = l + r;
			}else {
				int r = compute(i+1, end, 1- type, tokens);
				v = l -r;
			}

			if( COMP[type].compare(v, ret) > 0){
				ret = v;
			}
		}
		return cache[start][end][type] = ret;
	}


	private static int solve(String[] tokens){

		for(int[][] cell : cache){
			for(int[]  array : cell){
				Arrays.fill(array, Integer.MIN_VALUE);
			}
		}

		return compute(0, tokens.length, 0, tokens);
	}

	public static void main(String... args) throws IOException{

		Path p = Paths.get(System.getProperty("user.dir")+"/data/사칙연산.txt");
		BufferedReader rd = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(rd.readLine());
		while (testCases > 0){
			String[] tokens = rd.readLine().split("");
			int answer = Integer.parseInt(rd.readLine());

			System.out.println(String.format("%d, %d", solve(tokens), answer));
			testCases--;
		}

	}
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/사칙연산.kt
package com.example.cote.PGMRS

import java.io.*
import java.nio.file.*

object 사칙연산 {

    val MIN_ARRAY_LEGNTH = 3
    val MAX_ARRAY_LENGTH = 201
    val MIN_NUMBER_SIZE = 1000

    enum class Operator(val symbol: Char) {
        PLUS('+'),
        MINUS('-');

        val initValue: Int
            get() {
                return when (symbol) {
                    '+' -> Int.MIN_VALUE
                    '-' -> Int.MAX_VALUE
                    else -> 0
                }
            }

        fun compare(a: Int, b: Int): Int {
            return when (symbol) {
                '+' -> a.compareTo(b)
                '-' -> b.compareTo(a)
                else -> 0
            }
        }
    }

    val cache =
        Array(MAX_ARRAY_LENGTH + 1) {
            Array(MAX_ARRAY_LENGTH + 1) {
                IntArray(Operator.values().size) { Int.MIN_VALUE }
            }
        }

    fun compute(start: Int, end: Int, type: Operator, tokens: String): Int {
        var ret = cache[start][end][type.ordinal]

        if (ret != Int.MIN_VALUE) {
            return ret
        }

        if (start == end - 1) {
            return Character.getNumericValue(tokens[start])
        }

        ret = type.initValue

        for (i in start + 1 until end - 1 step 2) {
            val left = compute(start, i, type, tokens)
            var value = left

            when (tokens[i]) {
                Operator.PLUS.symbol -> {
                    val right = compute(i + 1, end, Operator.PLUS, tokens)
                    value += right
                }
                Operator.MINUS.symbol -> {
                    val right = compute(i + 1, end, Operator.MINUS, tokens)
                    value -= right
                }
                else -> {}
            }

            if (type.compare(value, ret) > 0) {
                ret = value
            }
        }

        cache[start][end][type.ordinal] = ret
        return ret
    }

    fun solve(tokens: String): Int {

        for (outer in cache) {
            for (inner in outer) {
                inner.fill(Int.MIN_VALUE)
            }
        }

        return compute(0, tokens.length, Operator.PLUS, tokens)
    }

    fun main() {
        val p = Paths.get(System.getProperty("user.dir") + "/data/사칙연산.txt")
        val reader = Files.newBufferedReader(p)

        repeat(reader.readLine().toInt()) {
            val tokens = reader.readLine()
            val answer = reader.readLine().toInt()
            println("$answer ${solve(tokens)}")
        }
    }
}

fun main() {
    사칙연산.main()
}
